qiskit/transpiler/passes/scheduling/instruction_alignment.py

```diff
@@ -79,117 +96,189 @@ class AlignMeasures(TransformationPass):
+            - The end time of current does not overlap with the start time of successor nodes.
+            - Compiler directives are not necessary satisfying the constraints.
+
+        Assumptions:
+
+            - Topological order and absolute time order of DAGOpNode are consistent.
+
+        Based on the configurations above, rescheduler pass takes following strategy.
+
+        1. Scan node from the begging, i.e. from left of the circuit. The rescheduler
+            calls ``node_start_time`` from the property set,
+            and retrieve the scheduled start time of current node.
+        2. If the start time of the node violates the alignment constraints,
+            the scheduler increases the start time until it satisfies the constraint.
+        3. Check overlap with successor nodes. If any overlap occurs, the rescheduler
+            recursively pushs the successor nodes backward towards the end of the wire.
+            Note that shifted location doesn't need to satisfy the constraints,
+            thus it will be a minimum delay to resolve the overlap with the ancestor node.
+        4. Repeat 1-3 until the node at the end of the wire. This will resolve
+            all misalignment wihtout inducing the overlap between the node.
+
+        Args:
+            dag: DAG circuit to be rescheduled with constraints.
+
+        Raises:
+            TranspilerError: Alignment is necessary but scheduling is not performed.
+        """
+        # Rescheduling is not necessary
+        if self.acquire_align == 1 and self.pulse_align == 1:
+            return
+
+        run_rescheduler = False
+
+        # Check delay durations
+        for delay_node in dag.op_nodes(Delay):
+            dur = delay_node.op.duration
+            if not (dur % self.acquire_align == 0 and dur % self.pulse_align == 0):
+                run_rescheduler = True
+                break
+
+        # Check custom gate durations
+        for inst_defs in dag.calibrations.values():
+            for caldef in inst_defs.values():
+                dur = caldef.duration
+                if not (dur % self.acquire_align == 0 and dur % self.pulse_align == 0):
+                    run_rescheduler = True
+                    break
+
+        if not run_rescheduler:
+            return
+
+        # Need scheduling for alignment. Error is caused when DAG circuit is not scheduled.
+        # This check should be placed here for backward compatibility.
+        # Now rescheduler pass is set to all preset pass managers,
+        # but scheduling is disabled by default.
+        # If this pass detect any violation risk, it will ask user to schedule the circuit.
+        if "node_start_time" not in self.property_set:
+            raise TranspilerError(
+                f"Input DAG {dag.name} likely need alignment but no scheduling is performed. "
+                "Set scheduling method or add explicit scheduling pass to the pass manager."
```

Comment:
```suggestion
        # Need scheduling for alignment. Error is caused when DAG circuit is not scheduled.
        # This check should be placed here for backward compatibility.
        # Now rescheduler pass is set to all preset pass managers,
        # but scheduling is disabled by default.
        # If this pass detect any violation risk, it will ask user to schedule the circuit.
        if "node_start_time" not in self.property_set:
            raise TranspilerError(
                f"Input DAG {dag.name} likely needs alignment but no scheduling is performed. "
                "Set scheduling method or add explicit scheduling pass to the pass manager."
```

Reply:
But I think if node_start_time is not available then this pass can't run and so should error. It's not about "circuit likely needs realignment", we just can't do anything without that information.

Reply:
This is intentional. Because of the overhead of scheduling, it is turned off by default (do nothing unless user set `scheduling_method="alap/asap"` in the transpiler arg). However, if user try to run T1 experiment from Qiskit Experiment (just an example), one will set interval in units of sec which usually doesn't satisfy these alignment. However, violation of acquire alignment is not detected on IBM Quantum backends, and it just returns erroneous results to users, i.e. random 0 or 1 count because frame is off. This is indeed hard for most of users to find the cause of this bug. Qiskit Experiment can handle this by setting default transpiling options, however, if it is a handmade program, one should know these constraints. 

So this pass can be split into two pieces
(1) validation; check if all duration of instructions are multiple of alignment values.
(2) do alignment if some violation risk found.

The error message says, validation of (1) fails, so there is a risk that the circuit returns unexpected result or may fail in execution. So "I want to reschedule but your pass manager doesn't activate the scheduling pass".

Reply:
This makes me think it's better to split the pass. Now we have two separate pass for (1) and (2).
https://github.com/Qiskit/qiskit-terra/pull/7762/commits/2e542aeffe76abdc6ff5714c9a4539d5e8b27d9b

Reply:
ok that makes sense.